crates/turbo-tasks-memory/src/task.rs

```diff
@@ -620,16 +898,96 @@ impl Task {
         schedule_task
     }
 
+    fn scopes_dirty_or_active<const LIKELY_ACTIVE: bool, const INCREMENT_UNFINISHED: bool>(
+        &self,
+        scopes: &TaskScopes,
+        backend: &MemoryBackend,
+    ) -> bool {
+        if !LIKELY_ACTIVE && INCREMENT_UNFINISHED {
+            let mut active = false;
+            let mut processed_scopes = 0;
+            for scope in scopes.iter() {
+                backend.with_scope(scope, |scope| {
+                    scope.increment_unfinished_tasks(backend);
+                    log_scope_update!("add unfinished task: {} -> {}", *scope.id, *self.id);
+                    if !active {
+                        processed_scopes += 1;
+                        let mut state = scope.state.lock();
+                        if state.is_active() {
+                            active = true;
+                        } else {
+                            state.add_dirty_task(self.id);
+                        }
+                    }
+                })
+            }
+            if active {
+                for scope in scopes.iter().take(processed_scopes) {
+                    backend.with_scope(scope, |scope| {
+                        let mut state = scope.state.lock();
+                        if !state.is_active() {
+                            state.remove_dirty_task(self.id);
+                        }
+                    })
+                }
+            }
+            return active;
+        }
+        if LIKELY_ACTIVE {
+            if INCREMENT_UNFINISHED {
+                let mut active = false;
+                for scope in scopes.iter() {
+                    backend.with_scope(scope, |scope| {
+                        scope.increment_unfinished_tasks(backend);
+                        active = active || scope.state.lock().is_active();
+                    })
+                }
+                if active {
+                    return true;
+                }
+            } else {
+                if scopes
+                    .iter()
+                    .any(|scope| backend.with_scope(scope, |scope| scope.state.lock().is_active()))
```

Comment:
You can move the `INCREMENT_UNFINISHED` to within to the closure if you use `try_fold` instead of `any`, avoiding the repeated logic here.

Something like:

```rust
let active = scopes
    .iter()
    .try_fold(false, |active, scope| {
        active = active || scope.state.lock().is_active();
        match (INCREMENT_UNFINISHED, active) {
            (true, _) => {
                scope.increment_unfinished_tasks(backend);
                ControlFlow::Continue(active)
            }
            (false, true) => ControlFlow::Break(()),
            (false, false) => ControlFlow::Continue(false),
        }
    }).unwrap_or(true);
```

Although now that I'm writing this out, I don't think it's very clear either.

Reply:
I'm not so sure about that. Since `INCREMENT_UNFINISHED` is a const, it'll likely get compiled down to either what `.any` would do anyway (`!INCREMENT_UNFINISHED`) or to a slightly more complicated fold.